kernel/comps/framebuffer/src/framebuffer.rs

```diff
@@ -0,0 +1,157 @@
+///
+/// # Notes
+///
+/// It is highly recommended to use a synchronization primitive, such as a `SpinLock`, to
+/// lock the framebuffer before performing any operation on it.
+/// Failing to properly synchronize access can result in corrupted framebuffer content
+/// or unspecified behavior during rendering.
+#[derive(Debug)]
+pub struct FrameBuffer {
+    io_mem: IoMem,
+    width: usize,
+    height: usize,
+    pixel_format: PixelFormat,
+}
+
+pub static FRAMEBUFFER: Once<Arc<FrameBuffer>> = Once::new();
+
+pub(crate) fn init() {
+    let Some(framebuffer_arg) = boot_info().framebuffer_arg else {
+        log::warn!("Framebuffer not found");
+        return;
+    };
+
+    if framebuffer_arg.address == 0 {
+        log::error!("Framebuffer address is zero");
+        return;
+    }
+
+    // FIXME: There are several pixel formats that have the same BPP. We lost the information
+    // during the boot phase, so here we guess the pixel format on a best effort basis.
+    let pixel_format = match framebuffer_arg.bpp {
+        8 => PixelFormat::Grayscale8,
+        16 => PixelFormat::Rgb565,
+        24 => PixelFormat::Rgb888,
+        32 => PixelFormat::Rgba,
+        _ => {
+            log::error!(
+                "Unsupported framebuffer pixel format: {} bpp",
+                framebuffer_arg.bpp
+            );
+            return;
+        }
+    };
+
+    let framebuffer = {
+        let fb_base = framebuffer_arg.address;
+        let fb_size = framebuffer_arg.width
+            * framebuffer_arg.height
+            * (framebuffer_arg.bpp / u8::BITS as usize);
+        let io_mem = IoMem::acquire(fb_base..fb_base + fb_size).unwrap();
+        FrameBuffer {
+            io_mem,
+            width: framebuffer_arg.width,
+            height: framebuffer_arg.height,
+            pixel_format,
+        }
+    };
+
+    framebuffer.clear();
+    FRAMEBUFFER.call_once(|| Arc::new(framebuffer));
+}
+
+impl FrameBuffer {
+    /// Returns the size of the framebuffer in bytes.
+    pub fn size(&self) -> usize {
+        self.io_mem.length()
+    }
+
+    /// Returns the width of the framebuffer in pixels.
+    pub fn width(&self) -> usize {
+        self.width
+    }
+
+    /// Returns the height of the framebuffer in pixels.
+    pub fn height(&self) -> usize {
+        self.height
+    }
+
+    /// Returns the pixel format of the framebuffer.
+    pub fn pixel_format(&self) -> PixelFormat {
+        self.pixel_format
+    }
+
+    /// Returns the rendered pixel.
+    pub fn render_pixel(&self, pixel: Pixel) -> RenderedPixel {
+        pixel.render(self.pixel_format)
+    }
+
+    /// Calculates the offset of a pixel at the specified position.
+    pub fn calc_offset(&self, x: usize, y: usize) -> PixelOffset {
+        PixelOffset {
+            fb: self,
+            offset: (y * self.width + x) * self.pixel_format.nbytes(),
+        }
+    }
+
+    fn check_offset(&self, offset: usize) -> Result<()> {
```

Comment:
`io_mem.write_bytes` will do the bound checking at its very first and guarantee there are no short writes. So I don't know why you need to repeat the bound checking here?

https://github.com/asterinas/asterinas/blob/1130933546fba3ffa6332b726d361e79f75eecfd/ostd/src/io/io_mem/mod.rs#L175-L181